zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/client/FederatedZkClient.java

```diff
@@ -477,10 +477,31 @@ public long getCreationTime(String path) {
     return getZkClient(path).getCreationTime(path);
   }
 
+  /**
+   * Executes ZkMulti on operations that are connected to the same Zk server.
+   * Will throw exception if any operation's server connection is different.
+   * @param ops
+   * @return
+   * @throws IllegalArgumentException
+   */
   @Override
   public List<OpResult> multi(Iterable<Op> ops) {
-    throwUnsupportedOperationException();
-    return null;
+    if (ops == null) {
+      throw new NullPointerException("ops must not be null.");
+    }
+    String opPath = null;
+    String opPathRealm = null;
+    for (Op op : ops) {
```

Comment:
From my understanding of code, in this for loop you want to filter ops who have same ZkRealm. Just wondering if there could be case where first few ops in given ops list won't have any realm but later ones have it. In that case, here even those first few ops will be sent for "multi" operation right? Should we use any lambda filter method which filters out all ops with same ZkRealm.

Reply:
I think an op must have a ZkRealm in order to be valid. Trying to get the path of an op with no ZkRealm would throw an error (in `TrieRoutingData.getMetadataStoreRealm()`) and so the multi operation won't succeed.